include/deal.II/differentiation/ad/ad_helpers.h

```diff
@@ -0,0 +1,683 @@
+       * standard computations. The operations that are performed with it are
+       * not recorded on the tape, and so should only be used when not in
+       * recording mode.
+       * @param[in] index The index in the vector of independent variables.
+       */
+      void
+      initialize_non_sensitive_independent_variable(
+        ad_type &          out,
+        const unsigned int index) const;
+
+      /**
+       * The number of independent variables that have been manipulated within a
+       * set of operations.
+       */
+      unsigned int
+      n_touched_independent_variables() const;
+      //@}
+
+      /**
+       * @name Dependent variables
+       */
+      //@{
+
+      /**
+       * The set of dependent variables $\mathbf{f}(\mathbf{X})$ of which the
+       * derivatives with respect to $\mathbf{X}$ will be computed.
+       *
+       * The gradients and Hessians of these dependent variables will be
+       * computed at the values $\mathbf{X}$ set with the set_sensitivity_value
+       * function.
+       *
+       * @note These are stored as an @p ad_type so that we can use them to
+       * compute function values and directional derivatives in the case that
+       * tapeless numbers are used
+       */
+      std::vector<ad_type> dependent_variables;
+
+      /**
+       * A list of dependent variables.
+       */
+      std::vector<bool> touched_dependent_variables;
+
+      /**
+       * Reset the boolean vector that indicates which independent variables
+       * that have been manipulated by the current set of operations.
+       * All entries in the vector are set to the value of the @p flag.
+       */
+      void
+      reset_touched_dependent_variables(const bool flag = false);
+
+      /**
+       * The number of dependent variables that have been registered.
+       */
+      unsigned int
+      n_touched_dependent_variables() const;
+
+      /**
+       * Register the definition of the index'th dependent variable
+       * $f(\mathbf{X})$.
+       *
+       * @param[in] func The recorded function that defines a dependent
+       * variable.
+       * @param[in] index The index of the entry in the global list of dependent
+       * variables that this function belongs to.
+       *
+       * @note Each dependent variable must only be registered once.
+       */
+      void
+      register_dependent_variable(const ad_type &    func,
+                                  const unsigned int index);
+      //@}
+
+    private:
+      /**
+       * A counter keeping track of the number of helpers in existence.
+       *
+       * This is only important information for when we use taped number types.
+       * As the tapes are stored in some global register, they exist independent
+       * of these helpers. However, it is assumed that when all helpers go
+       * out of scope then the tapes can be written over.
+       */
+      static dealii::types::global_dof_index n_helpers;
```

Comment:
is `global_dof_index` the right type here? I would think that a regular `unsigned int` is enough, or alternatively a `std::uint64_t`.